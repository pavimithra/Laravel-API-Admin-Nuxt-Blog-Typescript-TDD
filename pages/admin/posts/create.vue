<script setup lang="ts">
definePageMeta({
  middleware: ["auth"],
  layout: "admin",
  title: "Create Post",
});

useHead({
  title: "Create Post",
});
</script>

<template>
  <p>New Post</p>
</template>
