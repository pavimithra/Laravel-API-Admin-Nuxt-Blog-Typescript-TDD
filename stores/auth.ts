import { defineStore } from "pinia";
import axios from "axios";
import type { User, LoginPayload, RegisterPayload } from "@/types";

export const useAuthStore = defineStore("auth", () => {
  const user = ref<User | null>(null);
  const isAuthenticated = ref<boolean>(false);
  const loading = ref<boolean>(false);

  const login = async (payload: LoginPayload): Promise<void> => {
    await axios.post("/login", payload);
    await initUser();
  };

  const register = async (payload: RegisterPayload): Promise<void> => {
    await axios.post("/register", payload);
    await initUser();
  };

  const logout = async (): Promise<void> => {
    try {
      await axios.post("/logout");
      user.value = null;
      isAuthenticated.value = false;
    } catch (error) {
      throw new Error("Logout failed. Please try again.");
    } finally {
    }
  };

  const fetchUser = async (): Promise<User | null> => {
    try {
      const { data } = await axios.get("/user");
      return data;
    } catch (err) {
      return null;
    }
  };

  const initUser = async (): Promise<void> => {
    try {
      const userData = await fetchUser();
      if (userData) {
        user.value = userData;
        isAuthenticated.value = true;
      } else {
        isAuthenticated.value = false;
      }
    } catch (error) {
      console.error("Failed to initialize user:", error);
      user.value = null;
      isAuthenticated.value = false;
    } finally {
    }
  };

  return {
    user,
    isAuthenticated,
    loading,
    login,
    register,
    logout,
    initUser,
  };
});
