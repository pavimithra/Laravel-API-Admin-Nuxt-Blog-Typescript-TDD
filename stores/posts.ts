import { defineStore } from "pinia";
import axios from "axios";
import type {
  PostWithPagination,
  postSearchData,
  PostPayload,
  Post,
  Categories,
} from "@/types";

export const usePostStore = defineStore("post", () => {
  const posts: Ref<PostWithPagination | null> = ref(null);
  const selectedPosts: Ref<number[]> = ref([]);
  const isLoading: Ref<boolean> = ref(false);

  async function fetchPosts(
    page = 1,
    searchData: postSearchData
  ): Promise<void> {
    const { rowsPerPage, searchName, searchStatus } = searchData;

    isLoading.value = true; // Set loading state to true

    try {
      const response = await axios.get(`/posts?page=${page}`, {
        params: {
          rows_per_page: rowsPerPage,
          search_name: searchName,
          search_status: searchStatus,
        },
      });
      posts.value = response.data;
      selectedPosts.value = []; // Reset selectedPosts
    } finally {
      isLoading.value = false; // Set loading state to false after fetching
    }
  }

  async function storePost(postData: PostPayload): Promise<void> {
    const formData = new FormData();
    formData.append("title", postData.title);
    formData.append("slug", postData.slug);
    formData.append("description", postData.description);
    formData.append("content", postData.content);
    formData.append("category_id", postData.category_id);
    if (postData.image && postData.image.length > 0) {
      formData.append("image", postData.image[0].file);
    }

    await axios.post("/posts", formData, {
      headers: {
        "Content-Type": "multipart/form-data", // Ensures proper content type for FormData
      },
    });
  }

  async function showPost(postID: number): Promise<Post | null> {
    isLoading.value = true;
    try {
      const response = await axios.get(`/posts/${postID}`);
      return response.data as Post;
    } catch (error) {
      return null;
    } finally {
      isLoading.value = false; // Set loading state to false after fetching
    }
  }

  async function updatePost(
    postId: number,
    postData: PostPayload
  ): Promise<void> {
    const formData = new FormData();
    formData.append("_method", "PUT");
    formData.append("title", postData.title);
    formData.append("slug", postData.slug);
    formData.append("content", postData.content);
    formData.append("description", postData.description);
    formData.append("category_id", postData.category_id);
    const headers: Record<string, string> = {};
    if (postData.image && postData.image.length > 0) {
      formData.append("image", postData.image[0].file);
      headers["Content-Type"] = "multipart/form-data"; // Set content type only if image exists
    }
    await axios.post(`/posts/${postId}`, formData, {
      headers: headers, // Pass headers to the request
    });
  }

  async function deletePost(postId: number): Promise<void> {
    try {
      await axios.delete(`/posts/${postId}`);
    } catch (error) {
      console.error("Error deleting category:", error);
    }
  }

  async function fetchCategories(): Promise<Categories | null> {
    isLoading.value = true;
    try {
      const response = await axios.get("/posts/getCategories");
      return response.data as Categories;
    } catch (error) {
      return null;
    } finally {
      isLoading.value = false; // Set loading state to false after fetching
    }
  }

  async function performAction(actionType: string) {
    isLoading.value = true;
    try {
      const response = await axios.post("/posts/performAction", {
        actionType,
        selectedPosts: selectedPosts.value,
      });
      selectedPosts.value = [];
      await fetchPosts(1, {
        rowsPerPage: 5,
        searchName: "",
        searchStatus: "",
      });
      alert(response.data.message);
    } catch (error) {
      console.error("Error performing operation:", error);
    } finally {
      isLoading.value = false; // Set loading state to false after fetching
    }
  }

  return {
    posts: posts as ComputedRef<PostWithPagination | null>,
    selectedPosts,
    isLoading,
    fetchPosts,
    storePost,
    showPost,
    updatePost,
    deletePost,
    fetchCategories,
    performAction,
  };
});
