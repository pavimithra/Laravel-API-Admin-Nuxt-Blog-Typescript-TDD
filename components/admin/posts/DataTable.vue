<script setup lang="ts">
import { usePostStore } from "@/stores/posts";
// @ts-ignore
import { TailwindPagination } from "laravel-vue-pagination";
import {
  PencilSquareIcon,
  ArrowRightCircleIcon,
  TrashIcon,
} from "@heroicons/vue/24/outline";

const store = usePostStore();

const props = defineProps<{
  searchData: {
    rowsPerPage: number;
    searchName: string;
    searchStatus: string;
  };
}>();

const indeterminate = computed(
  () =>
    store.selectedPosts.length > 0 &&
    store.selectedPosts.length < (store.posts?.data?.length ?? 0)
);

const handleCheckboxChange = (event: Event) => {
  const target = event.target as HTMLInputElement;
  const postData = store.posts?.data;
  if (postData) {
    store.selectedPosts = target.checked ? postData.map((post) => post.id) : [];
  }
};

async function deletePost(postId: number) {
  try {
    await store.deletePost(postId);
    alert("Post Deleted Succesfully");
  } catch (err) {
    console.log("error");
  }
}
</script>

<template>
  <div class="relative">
    <div
      v-if="store.selectedPosts.length > 0"
      class="absolute left-14 top-0 flex h-12 items-center space-x-3 sm:left-12"
    >
      <button
        type="button"
        @click="store.performAction('publish')"
        data-testId="publish-all"
        class="inline-flex items-center rounded bg-white px-2 py-1 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50 disabled:cursor-not-allowed disabled:opacity-30 disabled:hover:bg-white"
      >
        Publish all
      </button>
      <button
        type="button"
        @click="store.performAction('delete')"
        data-testId="deletes-all"
        class="inline-flex items-center rounded bg-white px-2 py-1 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50 disabled:cursor-not-allowed disabled:opacity-30 disabled:hover:bg-white"
      >
        Delete all
      </button>
    </div>
    <table
      class="min-w-full table-fixed divide-y divide-gray-300 border border-gray-200 shadow-md"
    >
      <thead class="bg-gray-200">
        <tr>
          <th scope="col" class="relative px-7 sm:w-12 sm:px-6">
            <input
              v-if="
                store.isLoading ||
                (store.posts?.data && store.posts.data.length === 0)
              "
              type="checkbox"
              disabled
              class="absolute left-4 top-1/2 -mt-2 h-4 w-4 rounded border-gray-300 text-green-600 focus:ring-green-600"
            />
            <input
              v-else-if="store.posts?.data && store.posts.data.length > 0"
              type="checkbox"
              data-testId="checkbox-all"
              class="absolute left-4 top-1/2 -mt-2 h-4 w-4 rounded border-gray-300 text-green-600 focus:ring-green-600"
              :checked="
                indeterminate ||
                store.selectedPosts.length === store.posts?.data?.length
              "
              :indeterminate="indeterminate"
              @change="handleCheckboxChange"
            />
          </th>
          <th
            scope="col"
            class="min-w-[12rem] py-3.5 pr-3 text-left text-sm font-semibold text-gray-900"
          >
            Title
          </th>
          <th
            scope="col"
            class="px-3 py-3.5 text-left text-sm font-semibold text-gray-900 text-wrap"
          >
            Category
          </th>

          <th
            scope="col"
            class="px-3 py-3.5 text-left text-sm font-semibold text-gray-900"
          >
            Status
          </th>
          <th scope="col" class="relative py-3.5 pl-3 pr-4 sm:pr-3">
            <span class="sr-only">Edit</span>
          </th>
        </tr>
      </thead>
      <tbody class="divide-y divide-gray-200 bg-white">
        <tr
          v-for="post in store.posts?.data"
          :key="post.slug"
          :class="[store.selectedPosts.includes(post.id) && 'bg-gray-50']"
          v-if="!store.isLoading"
        >
          <td class="relative px-7 sm:w-12 sm:px-6">
            <div
              v-if="store.selectedPosts.includes(post.id)"
              class="absolute inset-y-0 left-0 w-0.5 bg-green-700"
            ></div>
            <input
              type="checkbox"
              :data-testId="'checkbox-' + post.id"
              class="absolute left-4 top-1/2 -mt-2 h-4 w-4 rounded border-gray-300 text-green-600 focus:ring-green-600"
              :value="post.id"
              v-model="store.selectedPosts"
            />
          </td>
          <td
            :class="[
              'py-4 pr-3 text-sm font-medium',
              store.selectedPosts.includes(post.id)
                ? 'text-indigo-600'
                : 'text-gray-900',
            ]"
          >
            <div class="flex items-center">
              <div class="h-10 w-10 flex-shrink-0">
                <img class="h-10 w-10 rounded-full" :src="post.image" alt="" />
              </div>
              <div class="ml-2">
                <div class="font-medium text-gray-900">
                  {{ post.title }}
                </div>
                <div class="mt-1 text-gray-500">{{ post.slug }}</div>
              </div>
            </div>
          </td>
          <td class="whitespace-nowrap px-3 py-4 text-sm text-gray-500">
            {{ post.category_name }}
          </td>
          <td class="whitespace-nowrap px-3 py-5 text-sm text-gray-500">
            <span
              class="inline-flex items-center rounded-md px-2 py-1 text-xs font-medium ring-1 ring-inset"
              :class="{
                'bg-green-50 text-green-700 ring-green-600/20':
                  post.status === 'published',
                'bg-yellow-50 text-yellow-700 ring-yellow-600/20':
                  post.status === 'draft',
              }"
            >
              {{ post.status }}
            </span>
          </td>
          <td
            class="whitespace-nowrap py-4 pl-3 pr-4 text-right text-sm font-medium sm:pr-3"
          >
            <div class="flex gap-3">
              <NuxtLink :to="'/admin/posts/' + post.id + '/show'">
                <ArrowRightCircleIcon
                  class="h-6 w-6 text-gray-700"
                  aria-hidden="true"
                />
              </NuxtLink>
              <NuxtLink :to="'/admin/posts/' + post.id + '/edit'">
                <PencilSquareIcon
                  class="h-6 w-6 text-gray-700"
                  aria-hidden="true"
                />
              </NuxtLink>
              <a href="#" @click="deletePost(post.id)">
                <TrashIcon class="h-6 w-6 text-gray-700" aria-hidden="true" />
              </a>
            </div>
          </td>
        </tr>
        <tr v-else class="text-center font-bold">
          <td
            colspan="5"
            class="whitespace-nowrap px-3 py-4 text-sm text-gray-500"
          >
            <span class="inline-flex items-center font-semibold leading-6">
              Loading...
            </span>
          </td>
        </tr>
        <tr
          v-if="
            !store.isLoading &&
            store.posts?.data &&
            store.posts.data.length === 0
          "
          class="text-center font-bold"
        >
          <td colspan="5" class="py-4 text-gray-500">No posts Found</td>
        </tr>
        <tr
          v-if="
            store.posts?.meta &&
            !store.isLoading &&
            store.posts.data &&
            store.posts.data.length > 0
          "
          class="text-center font-bold"
        >
          <td
            colspan="2"
            class="whitespace-nowrap px-3 py-4 text-sm text-gray-500 text-left"
          >
            Page <b>{{ store.posts.meta.current_page }}</b> of
            <b>{{ store.posts.meta.last_page }}</b>
          </td>
          <td
            colspan="3"
            class="whitespace-nowrap px-3 py-4 text-sm text-gray-500 text-right"
          >
            <TailwindPagination
              :data="store.posts"
              @pagination-change-page="
                        (page: number) => store.fetchPosts(page, props.searchData)
                      "
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>
