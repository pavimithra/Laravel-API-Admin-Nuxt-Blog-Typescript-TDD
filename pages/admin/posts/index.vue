<script setup lang="ts">
import { usePostStore } from "@/stores/posts";
import debounce from "lodash/debounce";

definePageMeta({
  middleware: ["auth"],
  layout: "admin",
  title: "Posts",
});

useHead({
  title: "Posts",
});

const store = usePostStore();

const searchData = reactive({
  rowsPerPage: ref(5),
  searchName: ref(""),
  searchStatus: ref(""),
});

onMounted(async () => {
  await store.fetchPosts(1, searchData);
});

const debouncedSearch = debounce(() => {
  store.fetchPosts(1, searchData);
}, 300);

watch(searchData, debouncedSearch, { deep: true });
</script>

<template>
  <div class="px-4">
    <div class="md:flex md:items-center">
      <div class="flex flex-grow">
        <NuxtLink to="/admin/posts/create">
          <button
            type="button"
            class="block uppercase rounded-md bg-green-600 px-4 py-2 text-center text-sm font-semibold leading-6 text-white shadow-sm hover:bg-green-700 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
          >
            Create Post
          </button>
        </NuxtLink>
      </div>
      <div class="mt-2 md:ml-5 lg:ml-16 md:mt-0 md:flex">
        <input
          type="text"
          name="name"
          id="name"
          data-testId="search-name"
          v-model="searchData.searchName"
          class="border border-zinc-500 rounded-lg mt-0 md:mt-2 lg:mt-0 text-base text-zinc-700 placeholder:text-zinc-400 bg-white focus-within:ring-1 focus-within:!ring-green-500 focus-within:!border-green-500"
          placeholder="Search By Name"
        />
        <div class="md:ml-4 mt-2 lg:mt-0">
          <label>
            <select
              id="searchStatus"
              data-testId="search-status"
              v-model="searchData.searchStatus"
              class="border border-zinc-500 rounded-lg text-base text-zinc-700 bg-white focus-within:ring-1 focus-within:!ring-green-500 focus-within:!border-green-500"
            >
              <option value="">Search Status</option>
              <option value="draft">Draft</option>
              <option value="published">Published</option>
            </select>
          </label>
        </div>
        <div class="md:ml-4 mt-2 lg:mt-0">
          <label>
            <select
              data-testId="rows-per-page"
              id="rowsPerPage"
              v-model="searchData.rowsPerPage"
              class="border border-zinc-500 rounded-lg text-base text-zinc-700 bg-white focus-within:ring-1 focus-within:!ring-green-500 focus-within:!border-green-500"
            >
              <option value="5">5</option>
              <option value="10">10</option>
              <option value="15">15</option>
              <option value="20">20</option>
              <option value="25">25</option>
            </select>
          </label>
        </div>
      </div>
    </div>

    <div class="mt-8 flow-root">
      <div class="-mx-4 -my-2 overflow-x-auto sm:-mx-6 lg:-mx-8">
        <div class="inline-block min-w-full py-2 align-middle sm:px-6 lg:px-8">
          <AdminPostsDataTable :searchData="searchData" />
        </div>
      </div>
    </div>
  </div>
</template>
