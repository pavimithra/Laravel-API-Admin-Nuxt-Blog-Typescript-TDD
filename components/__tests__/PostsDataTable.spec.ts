// @vitest-environment nuxt
import { describe, test, expect } from "vitest";
import { createTestingPinia } from "@pinia/testing";
import { mount, flushPromises } from "@vue/test-utils";
import { usePostStore } from "@/stores/posts";
import DataTable from "../admin/posts/DataTable.vue";
// @ts-ignore
import { TailwindPagination } from "laravel-vue-pagination";

describe("Posts DataTable Testing with No Data", async () => {
  const wrapper = mount(DataTable, {
    global: {
      plugins: [
        createTestingPinia({
          initialState: {
            post: {
              posts: {
                data: [],
                links: {
                  first: "#",
                  last: "#",
                  prev: null,
                  next: null,
                },
                meta: {
                  current_page: 1,
                  from: null,
                  last_page: 1,
                  links: [
                    {
                      url: null,
                      label: "&laquo; Previous",
                      active: false,
                    },
                    {
                      url: "#",
                      label: "1",
                      active: true,
                    },
                    {
                      url: null,
                      label: "Next &raquo;",
                      active: false,
                    },
                  ],
                  path: "#",
                  per_page: 5,
                  to: null,
                  total: 0,
                },
              },
            },
          },
        }),
      ],
    },
    props: {
      searchData: {
        rowsPerPage: 5,
        searchName: "",
        searchStatus: "",
      },
    },
  });

  await flushPromises(); // Wait for asynchronous operations to complete

  test("Displays no Posts Found inside the tbody", () => {
    // Assert that the component renders correctly
    expect(wrapper.exists()).toBe(true);

    // Assert that the table is rendered
    const table = wrapper.find("table");
    expect(table.exists()).toBe(true);

    // Assert that the table headers are rendered
    const tableHeaders = table.findAll("thead th");
    expect(tableHeaders.length).toBe(5); // Ensure correct number of headers

    // Assert that the "No posts Found" message is displayed
    const noPostsMessage = wrapper.find('tbody > tr > td[colspan="5"]');
    expect(noPostsMessage.exists()).toBe(true);
    expect(noPostsMessage.text()).toContain("No posts Found");
  });
});

describe("Posts DataTable Testing with Sample Data for 2 Pages", async () => {
  const wrapper = mount(DataTable, {
    global: {
      plugins: [
        createTestingPinia({
          initialState: {
            post: {
              selectedPosts: [],
              posts: {
                data: [
                  {
                    id: 1,
                    title: "Post 1",
                    slug: "post-1",
                    description: "Post Description 1",
                    content: "Post Content 1",
                    status: "draft",
                    image: "image1.png",
                  },
                  {
                    id: 2,
                    title: "Post 2",
                    slug: "post-2",
                    description: "Post Description 2",
                    content: "Post Content 2",
                    status: "draft",
                    image: "image2.png",
                  },
                  {
                    id: 3,
                    title: "Post 3",
                    slug: "post-3",
                    description: "Post Description 3",
                    content: "Post Content 3",
                    status: "draft",
                    image: "image3.png",
                  },
                ],
                links: {
                  first: "#",
                  last: "#",
                  prev: null,
                  next: null,
                },
                meta: {
                  current_page: 1,
                  from: 1,
                  last_page: 1,
                  links: [
                    {
                      url: null,
                      label: "&laquo; Previous",
                      active: false,
                    },
                    {
                      url: "#",
                      label: "1",
                      active: true,
                    },
                    {
                      url: null,
                      label: "Next &raquo;",
                      active: false,
                    },
                  ],
                  path: "#",
                  per_page: 5,
                  to: 3,
                  total: 3,
                },
              },
            },
          },
        }),
      ],
    },
    props: {
      searchData: {
        rowsPerPage: 5,
        searchName: "",
        searchStatus: "",
      },
    },
  });

  await flushPromises(); // Wait for asynchronous operations to complete

  const store = usePostStore();

  test("Displays Correct no of Datas and displays Page Details inside the tbody", () => {
    // Assert that the component renders correctly
    expect(wrapper.html()).toMatchSnapshot();

    // Additional assertions as needed
    expect(wrapper.find("table")).not.toBe(null); // Check if table exists
    expect(wrapper.findAll("tbody tr")).toHaveLength(4); // Check if 4 rows are rendered 3 data and 1 for no of pages

    // Assert that the component renders correctly
    expect(wrapper.exists()).toBe(true);

    // Assert that the table is rendered
    const table = wrapper.find("table");
    expect(table.exists()).toBe(true);

    // Assert that the table headers are rendered
    const tableHeaders = table.findAll("thead th");
    expect(tableHeaders.length).toBe(5); // Ensure correct number of headers

    // Assert that the "Page 1 of 1" page no is displayed
    const noPostsMessage = wrapper.find('tbody > tr > td[colspan="2"]');
    expect(noPostsMessage.exists()).toBe(true);
    expect(noPostsMessage.text()).toContain("Page 1 of 1");

    // Check if TailwindPagination component is rendered inside MyComponent
    expect(wrapper.findComponent(TailwindPagination).exists()).toBe(true);
  });

  test("test checkbox adds record ids to the selected posts ref correctly", async () => {
    // Assert that the buttons are not rendered
    expect(wrapper.find('[data-testId="publish-all"]').exists()).toBe(false);
    expect(wrapper.find('[data-testId="deletes-all"]').exists()).toBe(false);

    // Simulate checkbox all change
    const checkboxAll = wrapper.find("input[data-testId=checkbox-all]");
    await checkboxAll.trigger("click");
    await checkboxAll.trigger("change");

    // Assert that the selected posts are updated in the store
    expect(store.selectedPosts).toEqual([1, 2, 3]);

    // Assert that the buttons are rendered
    expect(wrapper.find('[data-testId="publish-all"]').exists()).toBe(true);
    expect(wrapper.find('[data-testId="deletes-all"]').exists()).toBe(true);

    store.selectedPosts = [];

    await flushPromises();

    // Simulate checkbox all change
    const checkboxID1 = wrapper.find("input[data-testId=checkbox-1]");
    await checkboxID1.trigger("click");
    await checkboxID1.trigger("change");

    // Assert that the selected posts are updated in the store
    expect(store.selectedPosts).toEqual([1]);

    // Simulate checkbox all change
    const checkboxID2 = wrapper.find("input[data-testId=checkbox-2]");
    await checkboxID2.trigger("click");
    await checkboxID2.trigger("change");

    // Assert that the selected posts are updated in the store
    expect(store.selectedPosts).toEqual([1, 2]);
  });
});
