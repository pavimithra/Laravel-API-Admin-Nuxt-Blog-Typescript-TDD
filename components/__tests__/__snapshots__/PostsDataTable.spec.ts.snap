// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Posts DataTable Testing with Sample Data for 2 Pages > Displays Correct no of Datas and displays Page Details inside the tbody 1`] = `
"<div class="relative">
  <!--v-if-->
  <table class="min-w-full table-fixed divide-y divide-gray-300 border border-gray-200 shadow-md">
    <thead class="bg-gray-200">
      <tr>
        <th scope="col" class="relative px-7 sm:w-12 sm:px-6"><input type="checkbox" data-testid="checkbox-all" class="absolute left-4 top-1/2 -mt-2 h-4 w-4 rounded border-gray-300 text-green-600 focus:ring-green-600"></th>
        <th scope="col" class="min-w-[12rem] py-3.5 pr-3 text-left text-sm font-semibold text-gray-900"> Title </th>
        <th scope="col" class="px-3 py-3.5 text-left text-sm font-semibold text-gray-900 text-wrap"> Category </th>
        <th scope="col" class="px-3 py-3.5 text-left text-sm font-semibold text-gray-900"> Status </th>
        <th scope="col" class="relative py-3.5 pl-3 pr-4 sm:pr-3"><span class="sr-only">Edit</span></th>
      </tr>
    </thead>
    <tbody class="divide-y divide-gray-200 bg-white">
      <tr class="">
        <td class="relative px-7 sm:w-12 sm:px-6">
          <!--v-if--><input type="checkbox" data-testid="checkbox-1" class="absolute left-4 top-1/2 -mt-2 h-4 w-4 rounded border-gray-300 text-green-600 focus:ring-green-600" value="1">
        </td>
        <td class="py-4 pr-3 text-sm font-medium text-gray-900">
          <div class="flex items-center">
            <div class="h-10 w-10 flex-shrink-0"><img class="h-10 w-10 rounded-full" src="image1.png" alt=""></div>
            <div class="ml-2">
              <div class="font-medium text-gray-900">Post 1</div>
              <div class="mt-1 text-gray-500">post-1</div>
            </div>
          </div>
        </td>
        <td class="whitespace-nowrap px-3 py-4 text-sm text-gray-500"></td>
        <td class="whitespace-nowrap px-3 py-5 text-sm text-gray-500"><span class="inline-flex items-center rounded-md px-2 py-1 text-xs font-medium ring-1 ring-inset bg-yellow-50 text-yellow-700 ring-yellow-600/20">draft</span></td>
        <td class="whitespace-nowrap py-4 pl-3 pr-4 text-right text-sm font-medium sm:pr-3">
          <div class="flex gap-3">
            <routerlink to="/admin/posts/1/show"></routerlink>
            <routerlink to="/admin/posts/1/edit"></routerlink><a href="#"><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" aria-hidden="true" data-slot="icon" class="h-6 w-6 text-gray-700">
                <path stroke-linecap="round" stroke-linejoin="round" d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"></path>
              </svg></a>
          </div>
        </td>
      </tr>
      <tr class="">
        <td class="relative px-7 sm:w-12 sm:px-6">
          <!--v-if--><input type="checkbox" data-testid="checkbox-2" class="absolute left-4 top-1/2 -mt-2 h-4 w-4 rounded border-gray-300 text-green-600 focus:ring-green-600" value="2">
        </td>
        <td class="py-4 pr-3 text-sm font-medium text-gray-900">
          <div class="flex items-center">
            <div class="h-10 w-10 flex-shrink-0"><img class="h-10 w-10 rounded-full" src="image2.png" alt=""></div>
            <div class="ml-2">
              <div class="font-medium text-gray-900">Post 2</div>
              <div class="mt-1 text-gray-500">post-2</div>
            </div>
          </div>
        </td>
        <td class="whitespace-nowrap px-3 py-4 text-sm text-gray-500"></td>
        <td class="whitespace-nowrap px-3 py-5 text-sm text-gray-500"><span class="inline-flex items-center rounded-md px-2 py-1 text-xs font-medium ring-1 ring-inset bg-yellow-50 text-yellow-700 ring-yellow-600/20">draft</span></td>
        <td class="whitespace-nowrap py-4 pl-3 pr-4 text-right text-sm font-medium sm:pr-3">
          <div class="flex gap-3">
            <routerlink to="/admin/posts/2/show"></routerlink>
            <routerlink to="/admin/posts/2/edit"></routerlink><a href="#"><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" aria-hidden="true" data-slot="icon" class="h-6 w-6 text-gray-700">
                <path stroke-linecap="round" stroke-linejoin="round" d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"></path>
              </svg></a>
          </div>
        </td>
      </tr>
      <tr class="">
        <td class="relative px-7 sm:w-12 sm:px-6">
          <!--v-if--><input type="checkbox" data-testid="checkbox-3" class="absolute left-4 top-1/2 -mt-2 h-4 w-4 rounded border-gray-300 text-green-600 focus:ring-green-600" value="3">
        </td>
        <td class="py-4 pr-3 text-sm font-medium text-gray-900">
          <div class="flex items-center">
            <div class="h-10 w-10 flex-shrink-0"><img class="h-10 w-10 rounded-full" src="image3.png" alt=""></div>
            <div class="ml-2">
              <div class="font-medium text-gray-900">Post 3</div>
              <div class="mt-1 text-gray-500">post-3</div>
            </div>
          </div>
        </td>
        <td class="whitespace-nowrap px-3 py-4 text-sm text-gray-500"></td>
        <td class="whitespace-nowrap px-3 py-5 text-sm text-gray-500"><span class="inline-flex items-center rounded-md px-2 py-1 text-xs font-medium ring-1 ring-inset bg-yellow-50 text-yellow-700 ring-yellow-600/20">draft</span></td>
        <td class="whitespace-nowrap py-4 pl-3 pr-4 text-right text-sm font-medium sm:pr-3">
          <div class="flex gap-3">
            <routerlink to="/admin/posts/3/show"></routerlink>
            <routerlink to="/admin/posts/3/edit"></routerlink><a href="#"><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" aria-hidden="true" data-slot="icon" class="h-6 w-6 text-gray-700">
                <path stroke-linecap="round" stroke-linejoin="round" d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"></path>
              </svg></a>
          </div>
        </td>
      </tr>
      <!--v-if-->
      <tr class="text-center font-bold">
        <td colspan="2" class="whitespace-nowrap px-3 py-4 text-sm text-gray-500 text-left"> Page <b>1</b> of <b>1</b></td>
        <td colspan="3" class="whitespace-nowrap px-3 py-4 text-sm text-gray-500 text-right">
          <!---->
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;
